### ISSUE TITLE
Feature request: Custom Text for Mini Cart (Menu Cart Widget) when Cart Empty

### ISSUE THREAD

**Issue by datverse** [elementor/elementor#16812]
Hi, 
Can I request a new feature for Menu Cart Widget with Type is "Mini Cart".
When Mini Cart is Empty it display the message is: "No products in the cart." 
Can you add a feature to custom this text?
Thanks

**Comment by Barns-be** [comment]
I use this workaround in Custom CSS:
.woocommerce-mini-cart__empty-message {
	visibility:hidden;
	position:relative;
}
.woocommerce-mini-cart__empty-message:after {
    visibility: visible;
    position: absolute;
    top: 0;
    left: 0;
    content: "The text you want it to be.";
}

**Comment by datverse** [comment]
Hi @Barns-be,
Thank for your reply. Right now that's the best way to fix this issue and I'm using it too. I think in the future we have feature custom message, label, text for Cart Widget when it empty items. This is helpful for Editor User (don't know how to "Hard code"). Not only with "Empty Message" but also with "Button Text" (The Price $0 before the Cart Icon).

<img width="890" alt="Screen Shot 2021-11-16 at 10 57 07 PM" src="https://user-images.githubusercontent.com/2121900/142020472-29d8f70a-45a4-453b-bc67-1a428d841a29.png">


Thanks

**Comment by juanisidoro** [comment]
> I use this workaround in Custom CSS: .woocommerce-mini-cart__empty-message { visibility:hidden; position:relative; } .woocommerce-mini-cart__empty-message:after { visibility: visible; position: absolute; top: 0; left: 0; content: "The text you want it to be."; }

Thank you very much!
This has been a quick fix ^^

**Comment by stefanullinger** [comment]
You must be kidding me... is this still not fixed?

**Comment by 79ho3ein** [comment]
This is a good idea. But it would be better if we could display a completely customized message to the users. Just like the custom empty cart feature that now exists for Elementor cart Widget.
Using this feature, we can even display the image, button and other widgets in the empty menu cart And that can be great.
Of course, @avivu  said in the beta pro 3.7 version that they are working on it. But now I don't know in what state it is and when it will be released.

**Comment by parsaselahi** [comment]
> I use this workaround in Custom CSS: .woocommerce-mini-cart__empty-message { visibility:hidden; position:relative; } .woocommerce-mini-cart__empty-message:after { visibility: visible; position: absolute; top: 0; left: 0; content: "The text you want it to be."; }

thanks to you, this one I developed is a bit better:

.woocommerce-mini-cart__empty-message {
visibility:hidden;
position:relative;
}
.woocommerce-mini-cart__empty-message:after {
visibility: visible;
white-space: pre;
position: absolute;
top: 0;
left: 0;
content: "The text you want it to be.";
}

the difference is in "white-space: pre;" witch make the content go to next line. and be a cleaner view.

**Comment by 79ho3ein** [comment]
Hi @parsaselahi , Thanks for the great code.
I think that for the menu cart widget, just being able to edit the text is not enough. Rather, it is better to be able to display a custom template instead of text. For example, you can write your own custom text and place an image, icon, or even a button below it. Of course, Elementor developers said in previous versions that they are working on it.

Hi @avivu , I apologize for tagging. Is there any news about the custom empty cart feature for the menu cart?

**Comment by avivu** [comment]
Hi @79ho3ein,
Sorry, but not at this point. We had some issues with implementing a custom template inside the Menu Cart widget, and it's still a WIP. We want to ensure we are releasing a stable feature.

**Comment by 79ho3ein** [comment]
It is true. Thank you @avivu

**Comment by datverse** [comment]
Hi @avivu , current we only need custom text (Useful when use on multi-language site). Can you update this first, we don't need content template for Cart Empty.

[Please view the screenshot](https://user-images.githubusercontent.com/2121900/142020472-29d8f70a-45a4-453b-bc67-1a428d841a29.png)

Thanks

**Comment by matintzh** [comment]
Hi guys , so I was searching the internet for a better fix without using the CSS , so here is what I found:
**1. if you are not using elementor**
go to your woocommerce plugin folder , Template > files > minicart.php 
search the massage text and change it to what you want , normally its like this :
`<p class="woocommerce-mini-cart__empty-message"><?php esc_html_e( 'No products in the cart.', 'woocommerce' ); ?></p>`

**2. if you are using elementor**
go to elementor pro plugin folder , elementor-pro > modules > woocommerce > wc-templates > cart
again search the massage and change it 
`<div class="woocommerce-mini-cart__empty-message"><?php esc_attr_e( 'No products in the cart.', 'elementor-pro' ); ?></div>`

hope this helps :)

**Comment by MattBri1992** [comment]
> Hi guys , so I was searching the internet for a better fix without using the CSS , so here is what I found: **1. if you are not using elementor** go to your woocommerce plugin folder , Template > files > minicart.php search the massage text and change it to what you want , normally its like this : `<p class="woocommerce-mini-cart__empty-message"><?php esc_html_e( 'No products in the cart.', 'woocommerce' ); ?></p>`
> 
> **2. if you are using elementor** go to elementor pro plugin folder , elementor-pro > modules > woocommerce > wc-templates > cart again search the massage and change it `<div class="woocommerce-mini-cart__empty-message"><?php esc_attr_e( 'No products in the cart.', 'elementor-pro' ); ?></div>`
> 
> hope this helps :)

Hi there thanks a lot for your contribution. However I tried to change it with Elementor Pro the way you suggested but somehow the text is not changing at all still keeps on saying "no products in basket" Is there anything more we can try? thanks a lot!

**Comment by rafamuz** [comment]
> Hi guys , so I was searching the internet for a better fix without using the CSS , so here is what I found: **1. if you are not using elementor** go to your woocommerce plugin folder , Template > files > minicart.php search the massage text and change it to what you want , normally its like this : `<p class="woocommerce-mini-cart__empty-message"><?php esc_html_e( 'No products in the cart.', 'woocommerce' ); ?></p>`
> 
> **2. if you are using elementor** go to elementor pro plugin folder , elementor-pro > modules > woocommerce > wc-templates > cart again search the massage and change it `<div class="woocommerce-mini-cart__empty-message"><?php esc_attr_e( 'No products in the cart.', 'elementor-pro' ); ?></div>`
> 
> hope this helps :)

the problem is when woocommerce updates to a newer version, Will it not be overwritten?

**Comment by rafamuz** [comment]
the problem is still happening. so simple solution for Elementor TEAM to fix, but nothing for years! please!!! common!

**Comment by dnajcz** [comment]
> still not fixed? haha
> 
> add to functions.php The effect may not be immediate. You may need to restart your browser. Because this text is cached by something.
> 
> `function tb_change_text( $translated_text ) { if ( $translated_text == 'No products in the cart.') { $translated_text = 'your message here'; } return $translated_text; } add_filter( 'gettext', 'tb_change_text', 20 );`

Yes, it is working. I would make it a little better for Elementor Pro users:
`add_filter( 'gettext', 'woocommerce_my_translate_text', 20, 3 );
function woocommerce_my_translate_text( $translated, $text, $domain ) {
	if ( $domain === 'elementor-pro' ) {
		$translated = str_ireplace('No products in the cart.', 'Košík je prázdný.', $translated);
	}
	return $translated;
}`
**NOTICE!** After this you need to add some product to the cart and remove it again so cache is updated.

**Comment by CreativeNomadUK** [comment]
I recently encountered this but I wanted to customize the empty cart so it can be populated with additional content and CTA's. It took awhile but I have it working. I created a new section in elementor with the content I wanted to use in the empty mini cart like some top selling products people can browse and a button that goes to the main store page. And then a function which can be added either in functions.php or code snippets:

Just remember to change this on line 2 - [elementor-template id="YOUR_ID_HERE"]

```
add_action('wp_footer', function() {
    $elementor_output = do_shortcode('[elementor-template id="YOUR_ID_HERE"]');
    $elementor_output = str_replace(array("\n", "\r"), '', $elementor_output);
    $elementor_output = addslashes($elementor_output); 
    ?>
<script type="text/javascript">
jQuery(function($){
    function injectMiniCartElementor() {
        var $miniCart = $('.widget_shopping_cart_content');

        if ($miniCart.length) {
            if ($miniCart.find('.woocommerce-mini-cart__empty-message').length) {
                // Remove previous injection
                $miniCart.find('.custom-mini-cart-elementor').remove();

                // Inject Elementor template
                $miniCart.append('<div class="custom-mini-cart-elementor"><?php echo $elementor_output; ?></div>');
            }
        }
    }

    // Run on page load
    injectMiniCartElementor();

    // Run after mini cart AJAX refresh
    $(document.body).on('wc_fragments_refreshed wc_fragments_loaded', function() {
        injectMiniCartElementor();
    });
});
</script>
```
<?php });
```
